### ISSUE TITLE
Module cannot both use completion widgets and other completion mechanisms

### ISSUE THREAD

**Issue by jcassee** [sorin-ionescu/prezto#1778]
### Description

As also mentioned in #1659, the completion module enables the completion system, but also initializes the completion widgets. This means:
* Modules loaded before completion that use `complete` for their completions will have their definitions cleared.
* Modules loaded after completion will not have their completion widgets initialized.

I have a module that uses both (because of external completion scripts), which causes some completions to be lost.

### Expected behavior

To be able to use both `complete` and completion widgets in the same module.

I believe completion widgets of modules loaded later could be included if the collection of `fpath` elements is a separate step before the execution of the `init.zsh` files, so `fpath` is already complete when the completion module is loaded. However, this conflicts with the removal of the entry from `fpath` if the module fails to load.

Alternatively, the completion module could be separated into two parts, one of which would be loaded early, and the other (`compinit`) late.

My current workaround is to call `compinit` again after all modules were loaded, but this seems wasteful.

### Actual behavior

Either `complete` or completion widgets work.

### Steps to Reproduce

1. Create a module containing both a completion for terraform and a completion widget.
  * init.sh:
    ```
    complete -o nospace -C /home/joost/.local/bin/terraform terraform`
    ```
  * functions/_example:
    ```
    #compdef example
    
    function _example {
      _arguments '1: :(foo bar)'
    }
    ```

2. Either include the module before or after the completion module. Either the terraform or example completion will not work.

### Versions

  - Prezto commit: 7c94154dda2dfcf02515f1caf1e9dfe6d80b0350
  - ZSH version: zsh 5.7.1 (x86_64-ubuntu-linux-gnu)
  - OS information: Ubuntu 19.10

**Comment by srijanshetty** [comment]
I agree with the changes in principle, I have a custom module and I have resorted to calling compinit in the same in order to enable completions.

**Comment by belak** [comment]
Huh, this is really interesting. I wonder if it would make sense to start adding some sort of prezto hooks, similar to zsh-add-hook... do you have any other ideas on how to call compinit later in the startup process?

**Comment by jcassee** [comment]
@belak As an outside contributor, I'm not inclined to suggest new framework-wide features. A hook for deferred execution sounds like it would work, but a simpler solution could be to split the call to `compinit` off into its own small module that is placed last in the modules list. Order is already important anyway. What would you like better?

**Comment by srijanshetty** [comment]
@belak, @jcassee I've faced the same issues and I use the workaround of deleting compinit cache and calling compinit everytime I add a new completion.

@jcassee's solution of splitting compinit is easier to implement but also easier to forget for the end user. I personally think that a prezto hook might be more suitable for this case, but can't think of any other use case of a custom prezto hook off the top of my head.
